pkg/migrate/cstor/volume.go

```diff
@@ -0,0 +1,902 @@
+			if err != nil {
+				return errors.Wrapf(err, "failed to retain pv {%s}", v.PVName)
+			}
+		}
+		err = v.updateStorageClass(pvObj.Name, pvObj.Spec.StorageClassName)
+		if err != nil {
+			return errors.Wrapf(err, "failed to update storageclass {%s}", pvObj.Spec.StorageClassName)
+		}
+		pvcObj, err = v.migratePVC(pvObj)
+		if err != nil {
+			return errors.Wrapf(err, "failed to migrate pvc to csi spec")
+		}
+	} else {
+		klog.Infof("PVC and storageclass already migrated to csi format")
+	}
+	v.StorageClass, err = v.KubeClientset.StorageV1().
+		StorageClasses().Get(*pvcObj.Spec.StorageClassName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	_, err = v.migratePV(pvcObj)
+	if err != nil {
+		return errors.Wrapf(err, "failed to migrate pv to csi spec")
+	}
+	err = v.removeOldTarget()
+	if err != nil {
+		return errors.Wrapf(err, "failed to remove old target deployment")
+	}
+	klog.Infof("Creating CVC to bound the volume and trigger CSI driver")
+	err = v.createCVC(v.PVName)
+	if err != nil {
+		return errors.Wrapf(err, "failed to create cvc")
+	}
+	err = v.validateMigratedVolume()
+	if err != nil {
+		return errors.Wrapf(err, "failed to validate migrated volume")
+	}
+	err = v.patchTargetPodAffinity()
+	if err != nil {
+		return errors.Wrap(err, "failed to patch target affinity")
+	}
+	err = v.cleanupOldResources()
+	if err != nil {
+		return errors.Wrapf(err, "failed to cleanup old volume resources")
+	}
+	return nil
+}
+
+func (v *VolumeMigrator) migratePVC(pvObj *corev1.PersistentVolume) (*corev1.PersistentVolumeClaim, error) {
+	klog.Infof("Generating equivalent CSI PVC")
+	pvcObj, recreateRequired, err := v.generateCSIPVC(pvObj.Name)
+	if err != nil {
+		return nil, err
+	}
+	if recreateRequired {
+		klog.Infof("Recreating equivalent CSI PVC")
+		pvcObj, err = v.RecreatePVC(pvcObj)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return pvcObj, nil
+}
+
+func (v *VolumeMigrator) migratePV(pvcObj *corev1.PersistentVolumeClaim) (*corev1.PersistentVolume, error) {
+	klog.Infof("Generating equivalent CSI PV")
+	pvObj, recreateRequired, err := v.generateCSIPV(pvcObj.Spec.VolumeName, pvcObj)
+	if err != nil {
+		return nil, err
+	}
+	if recreateRequired {
+		klog.Infof("Recreating equivalent CSI PV")
+		_, err = v.RecreatePV(pvObj)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return pvObj, nil
+}
+
+func (v *VolumeMigrator) generateCSIPVC(pvName string) (*corev1.PersistentVolumeClaim, bool, error) {
+	pvObj, err := v.KubeClientset.CoreV1().
+		PersistentVolumes().
+		Get(pvName, metav1.GetOptions{})
+	if err != nil {
+		return nil, false, err
+	}
+	pvcName := pvObj.Spec.ClaimRef.Name
+	pvcNamespace := pvObj.Spec.ClaimRef.Namespace
+	pvcObj, err := v.KubeClientset.CoreV1().PersistentVolumeClaims(pvcNamespace).
+		Get(pvcName, metav1.GetOptions{})
+	if err != nil {
+		if !k8serrors.IsNotFound(err) {
+			return nil, false, err
+		}
+	}
+	if pvcObj.Annotations["volume.beta.kubernetes.io/storage-provisioner"] != cstorCSIDriver {
```

Comment:
Code can crash here, right? If PVCObj is nil(After PVC deletion if pod restarts).

Reply:
If not found the Get function will return empty object.
```go
// Get takes name of the persistentVolumeClaim, and returns the corresponding persistentVolumeClaim object, and an error if there is any.
func (c *persistentVolumeClaims) Get(name string, options metav1.GetOptions) (result *v1.PersistentVolumeClaim, err error) {
	result = &v1.PersistentVolumeClaim{}
	err = c.client.Get().
		Namespace(c.ns).
		Resource("persistentvolumeclaims").
		Name(name).
		VersionedParams(&options, scheme.ParameterCodec).
		Do().
		Into(result)
	return
}
```
Also tested this case by 
 - setting PV to Retain and deleting the PVC.
 - then applied the job which succeeded.